lib/elixir_sense/providers/suggestion.ex

```diff
@@ -1,794 +1,96 @@
-        }
+  Adding suggestions:
 
-  @type protocol_function :: %{
-          type: :protocol_function,
-          name: String.t(),
-          arity: non_neg_integer,
-          args: String.t(),
-          origin: String.t(),
-          summary: String.t(),
-          spec: String.t(),
-          metadata: map
-        }
+      def my_reducer(hint, prefix, env, buffer_metadata, acc) do
+        suggestions = ...
+        {:cont, %{acc | result: acc.result ++ suggestions}}
+      end
 
-  @type func :: %{
-          type: :function | :macro,
-          name: String.t(),
-          arity: non_neg_integer,
-          args: String.t(),
-          origin: String.t(),
-          summary: String.t(),
-          spec: String.t(),
-          metadata: map
-        }
+  Defining the only set of suggestions to be provided:
 
-  @type mod :: %{
-          type: :module,
-          name: String.t(),
-          subtype: String.t(),
-          summary: String.t(),
-          metadata: map
-        }
+      def my_reducer(hint, prefix, env, buffer_metadata, acc) do
+        suggestions = ...
+        {:halt, %{acc | result: suggestions}}
+      end
 
-  @type param_option :: %{
-          type: :param_option,
-          name: String.t(),
-          origin: String.t(),
-          type_spec: String.t(),
-          doc: String.t(),
-          expanded_spec: String.t()
-        }
+  Defining a list of suggestions to be provided and allow an extra
+  limited set of additional reducers to run next:
 
-  @type type_spec :: %{
-          type: :type_spec,
-          name: String.t(),
-          arity: non_neg_integer,
-          origin: String.t(),
-          spec: String.t(),
-          doc: String.t(),
-          signature: String.t(),
-          metadata: map
-        }
+      def my_reducer(hint, prefix, env, buffer_metadata, acc) do
+        suggestions = ...
+        {:cont, %{acc | result: fields, reducers: [:populate_common, :variables]}}
+      end
+  """
 
-  @type hint :: %{
-          type: :hint,
-          value: String.t()
-        }
+  alias ElixirSense.Core.Metadata
+  alias ElixirSense.Core.State
+  alias ElixirSense.Providers.Suggestion.Reducers
 
   @type suggestion ::
-          attribute
-          | variable
-          | field
-          | return
-          | callback
-          | protocol_function
-          | func
-          | mod
-          | hint
-          | param_option
-          | type_spec
+          Reducers.Common.attribute()
+          | Reducers.Common.variable()
+          | Reducers.Struct.field()
+          | Reducers.Returns.return()
+          | Reducers.Callbacks.callback()
+          | Reducers.Protocol.protocol_function()
+          | Reducers.Common.func()
+          | Reducers.Common.mod()
+          | Reducers.Params.param_option()
+          | Reducers.TypeSpecs.type_spec()
+
+  @type acc :: %{result: [suggestion], reducers: [atom], context: map}
+
+  @reducers [
+    structs_fields: &Reducers.Struct.add_fields/5,
```

Comment:
Let's make it a `@behaviour`

Reply:
I was thinking about taking that road until I start working on the `Common` reducers. We have a `populate/5` and 5 others, which are very simple reducers: 

https://github.com/elixir-lsp/elixir_sense/blob/013f490e616c53057fb1b36cc955d117f3ad796e/lib/elixir_sense/providers/suggestion/reducers/common.ex#L84-L131

If we define a behaviour, It would require us to create 6 modules instead of one. So, since the golden rule usually is `data > function > module`, I left it as a function for now.

One thing that can easily change my mind is if we had to introduce new callbacks, for instance, instead of passing `reducers: [:populate_common, :variables]` [here](https://github.com/elixir-lsp/elixir_sense/blob/013f490e616c53057fb1b36cc955d117f3ad796e/lib/elixir_sense/providers/suggestion/reducers/struct.ex#L30), if we had a behaviour, we could define `name/0` and `dependencies/0` callback:

```elixir
def name() do
  :variables
end

def dependencies() do
  [:populate_common]
end
```

or maybe just one unified callback:

```elixir
def init() do
  {:variables, dependencies: [:populate_common]}
end
```

and then we could just use:

```elixir
  reducers: [:variables]
```

I think as soon as we start refactoring the reducers in order to optimize them, e.g. reusing data structures previously fetched by other reducers, we will need this. If you already have something in mind in that direction or you see other benefits, please let me know. I'd be glad to convert it into a behaviour.